src/gt4sd/training_pipelines/diffusion/core.py

```diff
@@ -0,0 +1,424 @@
+#
+# MIT License
+#
+# Copyright (c) 2022 GT4SD team
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in all
+# copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+#
+"""Diffusion training utilities. Code adapted from: https://github.com/huggingface/diffusers/blob/main/examples/unconditional_image_generation/train_unconditional.py"""
+
+import logging
+import os
+from dataclasses import dataclass, field
+from typing import Any, Dict
+
+import torch
+import torch.nn.functional as F
+from accelerate import Accelerator
+from datasets import load_dataset
+from diffusers import DDPMPipeline, DDPMScheduler, UNet2DModel
+from diffusers.optimization import get_scheduler
+from diffusers.training_utils import EMAModel
+from rdkit.Chem import Descriptors
+from torchvision.transforms import (
+    CenterCrop,
+    Compose,
+    InterpolationMode,
+    Normalize,
+    RandomHorizontalFlip,
+    Resize,
+    ToTensor,
+)
+from tqdm.auto import tqdm
+
+from ..core import TrainingPipeline, TrainingPipelineArguments
+
+logger = logging.getLogger(__name__)
+logger.addHandler(logging.NullHandler())
+
+# NOTE: This import is needed because importing torchvision before rdkit Descriptors
+# can cause segmentation faults.
+Descriptors
+
+
+class DiffusionTrainingPipeline(TrainingPipeline):
+    """Diffusion training pipelines."""
```

Comment:
Still waiting for the integration of diffusion models on molecules in diffusers. It will be probably easier to have two separate TrainingPipeline. What about `DiffusionTrainingPipelineVision`?

Reply:
Not bad, but to be consistent with the other pipelines it would be cool to keep as suffix `TrainingPipeline`.
Maybe `DiffusionForVisionTrainingPipeline`? or something along this line